fn.go

```diff
@@ -20,34 +31,182 @@ type Function struct {
 	log logging.Logger
 }
 
+const (
+	invalidFunctionFmt = "invalid function input: %s"
+	wrongTempErr       = "templates are required either inline or from filesystem path"
+	cannotGetErr       = "cannot get the function cd"
+	cannotParseErr     = "cannot parse the provided templates"
+)
+
+const (
+	AnnotationKeyCompositionResourceName = "crossplane.io/composition-resource-name"
+)
+
 // RunFunction runs the Function.
 func (f *Function) RunFunction(_ context.Context, req *fnv1beta1.RunFunctionRequest) (*fnv1beta1.RunFunctionResponse, error) {
 	f.log.Info("Running Function", "tag", req.GetMeta().GetTag())
 
-	// This creates a new response to the supplied request. Note that Functions
-	// are run in a pipeline! Other Functions may have run before this one. If
-	// they did, response.To will copy their desired state from req to rsp. Be
-	// sure to pass through any desired state your Function is not concerned
-	// with unmodified.
 	rsp := response.To(req, response.DefaultTTL)
 
-	// Input is supplied by the author of a Composition when they choose to run
-	// your Function. Input is arbitrary, except that it must be a KRM-like
-	// object. Supporting input is also optional - if you don't need to you can
-	// delete this, and delete the input directory.
 	in := &v1beta1.Input{}
 	if err := request.GetInput(req, in); err != nil {
-		response.Fatal(rsp, errors.Wrapf(err, "cannot get Function input from %T", req))
+		response.Fatal(rsp, errors.Wrapf(err, "cannot get Function cd from %T", req))
+		return rsp, nil
+	}
+
+	if !isValidInputSource(in) {
```

Comment:
Looks like `NewTemplateSourceGetter` is doing this same check, so we can remove this one.